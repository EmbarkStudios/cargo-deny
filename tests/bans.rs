use cargo_deny::{func_name, test_utils::*};

/// Covers issue <https://github.com/EmbarkStudios/cargo-deny/issues/184>
#[test]
fn cyclic_dependencies_do_not_cause_infinite_loop() {
    gather_bans(
        func_name!(),
        KrateGather::new("cyclic_dependencies"),
        Config::default(),
    );
}

#[test]
fn allow_wrappers() {
    let diags = gather_bans(
        func_name!(),
        KrateGather::new("allow_wrappers/maincrate"),
        r#"
[[deny]]
name = "dangerous-dep"
wrappers = ["safe-wrapper"]
"#,
    );

    insta::assert_json_snapshot!(diags);
}

#[test]
fn disallows_denied() {
    let diags = gather_bans(
        func_name!(),
        KrateGather::new("allow_wrappers/maincrate"),
        "deny = [{name = 'dangerous-dep'}]",
    );

    insta::assert_json_snapshot!(diags);
}

/// Validates that wildcard '*' dependencies can be detected and banned
#[test]
fn deny_wildcards() {
    let diags = gather_bans(
        func_name!(),
        KrateGather::new("wildcards/maincrate"),
        "wildcards = 'deny'",
    );

    insta::assert_json_snapshot!(diags);
}

/// Ensures that multiple versions are always deterministically sorted by
/// version number
/// See <https://github.com/EmbarkStudios/cargo-deny/issues/384>
#[test]
fn deterministic_duplicate_ordering() {
    let diags = gather_bans(
        func_name!(),
        KrateGather::new("duplicates"),
        "multiple-versions = 'deny'",
    );

    insta::assert_json_snapshot!(diags);
}
