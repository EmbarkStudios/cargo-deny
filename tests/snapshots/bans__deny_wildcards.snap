---
source: tests/bans.rs
expression: diags
---
[
  {
    "fields": {
      "code": "B006",
      "graphs": [
        {
          "Krate": {
            "name": "wildcards-test-crate",
            "version": "0.1.0"
          }
        }
      ],
      "labels": [
        {
          "column": 1,
          "line": 1,
          "message": "wildcard crate entry",
          "span": "ansi_term = '*'"
        }
      ],
      "message": "found 1 wildcard dependency for crate 'wildcards-test-crate'",
      "severity": "error"
    },
    "type": "diagnostic"
  },
  {
    "fields": {
      "code": "B006",
      "graphs": [
        {
          "Krate": {
            "name": "wildcards-test-dep",
            "version": "0.1.0"
          },
          "parents": [
            {
              "Krate": {
                "name": "wildcards-test-crate",
                "version": "0.1.0"
              }
            }
          ]
        }
      ],
      "labels": [
        {
          "column": 1,
          "line": 1,
          "message": "wildcard crate entry",
          "span": "itoa = '*'"
        }
      ],
      "message": "found 1 wildcard dependency for crate 'wildcards-test-dep'",
      "severity": "error"
    },
    "type": "diagnostic"
  }
]
